zeusd/src/devices/cpu/linux.rs

```diff
@@ -0,0 +1,219 @@
+
+        match fs::read_dir(&base_path) {
+            Ok(entries) => {
+                for entry in entries {
+                    if let Ok(entry) = entry {
+                        let path = entry.path();
+                        if path.is_dir() {
+                            if let Some(dir_name_str) = path.file_name() {
+                                let dir_name = dir_name_str.to_string_lossy();
+                                if dir_name.contains("intel-rapl") {
+                                    index_count += 1;
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+            Err(_) => {
+                tracing::error!("RAPL not available");
+            }
+        };
+        Ok(index_count)
+    }
+
+    fn get_available_fields(index: u32) -> Result<(PackageInfo, Option<PackageInfo>), ZeusdError> {
+        let base_path = PathBuf::from(format!("{}/intel-rapl:{}", RAPL_DIR, index));
+        let cpu_info = PackageInfo::new(&base_path, index)?;
+
+        let mut dram_info = None;
+        match fs::read_dir(&base_path) {
+            Ok(entries) => {
+                for entry in entries {
+                    if let Ok(entry) = entry {
+                        let path = entry.path();
+                        if path.is_dir() {
+                            if let Some(dir_name_str) = path.file_name() {
+                                let dir_name = dir_name_str.to_string_lossy();
+                                if dir_name.contains("intel-rapl") {
+                                    let subpackage_path = base_path.join(&*dir_name);
+                                    let subpackage_info =
+                                        PackageInfo::new(&subpackage_path, index)?;
+                                    if subpackage_info.name == "dram" {
+                                        dram_info = Some(subpackage_info);
+                                        break;
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+            Err(_) => {
+                return Err(ZeusdError::CpuInitializationError(
+                    index.try_into().unwrap(),
+                ))
+            }
+        };
+
+        Ok((cpu_info, dram_info))
+    }
+
+    fn get_cpu_energy(&self) -> anyhow::Result<u64, ZeusdError> {
+        let handle = self
+            .cpu_monitoring_task
+            .get_or_init(|| tokio::spawn(monitor_rapl(Arc::clone(&self.cpu))));
+        if handle.is_finished() {
+            return Err(ZeusdError::CpuManagementTaskTerminatedError(
+                self.cpu.index.try_into().unwrap(),
+            ));
+        }
+
+        let measurement = read_u64(&self.cpu.energy_uj_path)?;
+        let num_wraparounds = self.cpu.num_wraparounds.read().unwrap();
```

Comment:
You'll handle these unwraps soon, right?

Reply:
I meant to ask about this during our meeting. Would `map_err` such as this:
```
let num_wraparounds = self.cpu.num_wraparounds.read().map_err(|_| ZeusdError::CpuManagementTaskTerminatedError(self.cpu.index)).unwrap();

```

or converting `PosionError`s to `ZeusdError`
```
ZeusdError {...
    #[error("RWLockError: {0}")]
    RWLockError(String)
}

impl<T> From<PoisonError<T>> for ZeusdError {
    fn from(err: PoisonError<T>) -> Self {
        ZeusdError::RWLockError(format!("RWLockError: {}", err))
    }
}
```
be a better way of handling this error?

Reply:
I would just `map_err` and then use the `?` operator to return from the function. I also realize this is returning `anyhow::Result`, which isn't the best design.